File: leetcode/Remove_Duplicates_from_Sorted_List.java
import java.util.*;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null) return null;
        ListNode prev = head;
        ListNode cur = prev.next;
        while (cur != null) {
            if (prev.val != cur.val) {
                prev.next = cur;
                prev = prev.next;
            }
            cur = cur.next;
        }
        prev.next = null;
        return head;
    }
}

File: leetcode/Remove_Duplicates_from_Sorted_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *deleteDuplicates(ListNode *head) {
        // The input list is sorted.
        ListNode *pre = NULL;
        ListNode *cur = head;
        while (cur != NULL) {
            if (pre != NULL && pre->val == cur->val) {
                pre->next = cur->next;
                ListNode *deleting = cur;
                cur = cur->next;
                delete deleting;
            } else {
                pre = cur;
                cur = cur->next;
            }
        }
        return head;
    }
};
